src/pool/cli.c

```diff
@@ -1454,6 +1422,452 @@ dc_pool_query(tse_task_t *task)
 	return rc;
 }
 
+/*
+ * Is the cached pool map known to be stale? Must be called under
+ * pool->dp_map_lock.
+ */
+static bool
+map_known_stale(struct dc_pool *pool)
+{
+	unsigned int cached = pool_map_get_version(pool->dp_map);
+
+	D_ASSERTF(pool->dp_map_version_known >= cached, "%u >= %u\n",
+		  pool->dp_map_version_known, cached);
+
+	return (pool->dp_map_version_known > cached);
+}
+
+/*
+ * Arg and state of map_refresh
+ *
+ * mra_i is an index in the internal node array of a pool map. It is used to
+ * perform a round robin of the array starting from a random element.
+ */
+struct map_refresh_arg {
+	struct dc_pool	       *mra_pool;
+	bool			mra_passive;
+	unsigned int		mra_map_version;
+	int			mra_i;
+	struct d_backoff_seq	mra_backoff_seq;
+};
+
+/*
+ * When called repeatedly, this performs a round robin of the pool map rank
+ * array starting from a random index.
+ */
+static d_rank_t
+choose_map_refresh_rank(struct map_refresh_arg *arg)
+{
+	struct pool_domain     *nodes;
+	int			n;
+	int			i;
+	int			j;
+	int			k = -1;
+
+	n = pool_map_find_nodes(arg->mra_pool->dp_map, PO_COMP_ID_ALL, &nodes);
+	/* There must be at least one rank. */
+	D_ASSERTF(n > 0, "%d\n", n);
+
+	if (arg->mra_i == -1) {
+		/* Let i be a random integer in [0, n). */
+		i = ((double)rand() / RAND_MAX) * n;
```

Comment:
Good practice for a shared library would say that we don't use rand() or srand() to avoid any potential conflict with what the application expects, so the daos client should avoid calling either of these functions.  We do this in the fault injection code for example where we maintain our own bits and use nrand48() on them.

Reply:
@ashleypittman, could you point me to the srouce of the "good practice for a shared library", or elaborate, please? E.g., what "conflict" and what do applications expect? Thanks.

Reply:
There are many use-cases where people would want "random" numbers where the sequence is deterministic, and can be replicated if required, you can see this in the srand() and rand() APIs where a known seed will return a consistent sequence.

Essentially there's an assumption that the following code would be deterministic, although looking through our code that @kccain was referring to I think think this PR is a regression in this regard.

I've personally done this for testing where we use a random sequence for setting up groups/RPCs etc, and the test itself uses the time as a seed so it does different things every run, however it logs the seed and if there is a test failure you can replicate the specific test by using the seed as an input to the reproducer.

`#include <stdlib.h>
#include <stdio.h>

int main()
{
	srand(7);
	printf("%d\n", rand());
	printf("%d\n", rand());
	printf("%d\n", rand());
	daos_init();
	printf("%d\n", rand());
}
`

Reply:
I see. I had the same thought in the past; reality convinced me that the most practical way to achieve the deterministic sequence is for the test to use its own PRNG algorithm + its own instance of that algorithm.

Reply:
I agree, and any serious scientific application will do it's own thing with regard to random numbers anyway, but that's no reason for us to knowingly break programs that used the simplistic approach above.  For the server this is fine obviously, but as a shared library the onus is on us not to break 3rd party APIs like this.

It's easy enough for us to avoid, we simply need to use the functions listed here instead https://linux.die.net/man/3/drand48_r

Reply:
@ashleypittman, no objection, though I still struggle to view it as "breaking 3rd-party APIs". Let's create a new ticket for this problem, work out a new coding rule, as well as a way to communicate and enforce the new rule. Sounds OK to you?

Reply:
OK, just filed https://jira.hpdd.intel.com/browse/DAOS-8137.